SimTKcommon/src/ParallelExecutor.cpp

```diff
@@ -25,23 +25,36 @@
 #include "SimTKcommon/internal/ParallelExecutor.h"
 #include <pthread.h>
 
+#include <iostream>
+#include <string>
+#include <algorithm>
+
+using namespace std;
+
 namespace SimTK {
 
 static void* threadBody(void* args);
 
+ParallelExecutorImpl::ParallelExecutorImpl() : finished(false) {
+
+    // Set the maximum number of threads that we can use
+    SimTK_APIARGCHECK_ALWAYS(ParallelExecutor::getNumProcessors() > 0, "ParallelExecutorImpl", "ParallelExecutorImpl", "Number of threads detected is not positive.");
+    NUM_MAX_THREADS = ParallelExecutor::getNumProcessors();
+
+    pthread_mutex_init(&runLock, NULL);
+    pthread_cond_init(&runCondition, NULL);
+    pthread_cond_init(&waitCondition, NULL);
+
+}
 ParallelExecutorImpl::ParallelExecutorImpl(int numThreads) : finished(false) {
 
-    // Construct all the threading related objects we will need.
-    
+    // Set the maximum number of threads that we can use
     SimTK_APIARGCHECK_ALWAYS(numThreads > 0, "ParallelExecutorImpl", "ParallelExecutorImpl", "Number of threads must be positive.");
-    threads.resize(numThreads);
+    NUM_MAX_THREADS = numThreads;
+
     pthread_mutex_init(&runLock, NULL);
     pthread_cond_init(&runCondition, NULL);
     pthread_cond_init(&waitCondition, NULL);
-    for (int i = 0; i < numThreads; ++i) {
-        threadInfo.push_back(new ThreadInfo(i, this));
-        pthread_create(&threads[i], NULL, threadBody, threadInfo[i]);
-    }
 }
```

Comment:
> if all of the forces were non-parallel but the ParallelExecutor was constructed with default (empty) parameters it would spend time launching threads

Makes sense. Could you instead handle that by not creating a ParallelExecutor unless there is more than one thread? I don't feel strongly about it but it could be simpler to have either (a) no ParallelExecutor and everything runs on the main thread, or (b) a ParallelExecutor that creates the max number of threads on construction, then uses a subset of those as needed. The way it is is fine if you can be sure the container of threads makes sense upon reallocation. You can't verify that in the debugger because it could behave differently on different machines; it's not the behavior of `Array_` (same on all machines) but the behavior of copying or moving whatever object is being used to represent threads -- that takes looking at the code and making sure you know what's going on there. For example, we wouldn't be able to tell if the resizing created new threads because the old ones would just sit around doing nothing. `Array_` will std::move its contents during resize *if* those objects have move constructors, otherwise it is going to invoke copy constructors.